8357650: ThreadSnapshot to take snapshot of thread for thread dumps

alexmenkov opened the pull request:
This is first (hotspot) part of the update for `HotSpotDiagnosticMXBean.dumpThreads` and `jcmd Thread.dump_to_file` to include lock information in thread dumps (JDK-8356870).
The update has been split into parts to simplify reviewing.
The fix contains an implementation of `jdk.internal.vm.ThreadSnapshot` class to gather required information about a thread.
Second (dependent) part includes changes in `HotSpotDiagnosticMXBean.dumpThreads`/`jcmd Thread.dump_to_file`, spec updates and tests for the functionality.

Testing: new `HotSpotDiagnosticMXBean.dumpThreads`/`jcmd Thread.dump_to_file` functionality was tested in loom repo;
  sanity tier1 (this fix only)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8357650](https://bugs.openjdk.org/browse/JDK-8357650): ThreadSnapshot to take snapshot of thread for thread dumps (**Sub-task** - P4)


### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**)

### Contributors
 * Alan Bateman `<alanb@openjdk.org>`
 * Alex Menkov `<amenkov@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25425/head:pull/25425` \
`$ git checkout pull/25425`

Update a local copy of the PR: \
`$ git checkout pull/25425` \
`$ git pull https://git.openjdk.org/jdk.git pull/25425/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25425`

View PR using the GUI difftool: \
`$ git pr show -t 25425`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25425.diff">https://git.openjdk.org/jdk/pull/25425.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25425#issuecomment-2906082944)
</details>

alexmenkov commented:
/label remove core-libs

alexmenkov commented:
/contributor add alexmenkov
/contributor add alanb

alexmenkov commented:
/contributor add @alexmenkov

liach commented on src/java.base/share/classes/jdk/internal/vm/ThreadSnapshot.java:
```diff
@@ -0,0 +1,222 @@
+/**
+ * Represents a snapshot of information about a Thread.
+ */
+class ThreadSnapshot {
```
All class declarations here should be made final (except enum and record which are already implicitly final)

AlanBateman commented:
> 1. Did you consider putting this implementation in another file, say threadServices.hpp or even its own file? I'm asking because javaClasses seems to have become a dumping ground for code that calls from Java code into the JVM. It would be nice if we could put features into files that are named after the feature.

Alex can correct me but I think it started out here because of java_lang_Thread::async_get_stack_trace/GetStackTraceClosure. It has gone through a number of iterations and you are probably right that it would be better off in threadService or its own source file now.

> 2. Did you consider to put these OopHandle storages in something else than Universe::vm_globals()? We have, for example, ThreadService::_thread_service_storage.

That OopStorage is maintained in threadService.cpp so would be an option if the code moves.

alexmenkov commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1999,6 +2238,227 @@ oop java_lang_Thread::async_get_stack_trace(oop java_thread, TRAPS) {
   return trace();
 }
 
+class jdk_internal_vm_ThreadLock: AllStatic {
```
`jdk_internal_vm_ThreadLock` and `jdk_internal_vm_ThreadSnapshot` classes are "local", they are used  only for `dump_to_file` diagnostic command, so I decide to not pollute javaClasses.hpp and initialize them on demand (most other classes are initialized on VM startup or read from cds)

alexmenkov commented:
- moved implementation from javaClasses to threadServices;
- used ThreadService OopStorage instead of vm_global;
- fixed minimalVM build (Thread.dump_to_file command is not registered for builds without JVMTI);

dholmes-ora commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+
+  // StackTrace
+  InstanceKlass* ste_klass = vmClasses::StackTraceElement_klass();
+  assert(ste_klass != nullptr, "must be loaded in 1.4+");
```
```suggestion
  assert(ste_klass != nullptr, "must be loaded");
```
The 1.4+ is not relevant these days

dholmes-ora commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+  // StackTrace
+  InstanceKlass* ste_klass = vmClasses::StackTraceElement_klass();
+  assert(ste_klass != nullptr, "must be loaded in 1.4+");
+  if (ste_klass->should_be_initialized()) {
+    ste_klass->initialize(CHECK_NULL);
+  }
```
Is this actually necessary? Doesn't this klass always get initialized as part of VM initialization?

alexmenkov replied:
removed

dholmes-ora commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+
+  // call static StackTraceElement[] StackTraceElement.of(StackTraceElement[] stackTrace)
+  // to properly initialize STE.
+  {
```
Why the additional block scope?

dholmes-ora reviewed (commented):
Some initial drive-by comments ...

sspitsyn commented on src/hotspot/share/prims/jvm.cpp:
```diff
@@ -2961,6 +2961,15 @@ JVM_ENTRY(jobject, JVM_GetStackTrace(JNIEnv *env, jobject jthread))
+#if INCLUDE_JVMTI
+  oop snapshot = VMThreadSnapshot::get_thread_snapshot(jthread, THREAD);
+  return JNIHandles::make_local(THREAD, snapshot);
+#elif
```
Q: should it be `#else` instead of `#elif`?

alexmenkov replied:
Fixed

AlanBateman commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1872,7 +1872,7 @@ ByteSize java_lang_Thread::thread_id_offset() {
 
 oop java_lang_Thread::park_blocker(oop java_thread) {
-  return java_thread->obj_field(_park_blocker_offset);
+  return java_thread->obj_field_acquire(_park_blocker_offset);
```
Thread.parkBlocker is declared volatile but all access is done in opaque mode, so maybe this should MO_RELAXED?

In any case, I think we might have to drop the sampling of exclusiveOwnerThread for now. This can only be tested to see if it's the current thread, and only after a volatile-state of the state.

alexmenkov replied:
made access MO_RELAXED

alexmenkov commented on src/hotspot/share/prims/jvmtiThreadState.hpp:
```diff
@@ -77,7 +77,7 @@ class JvmtiEnvThreadStateIterator : public StackObj {
 // Virtual Thread Mount State Transition (VTMS transition) mechanism
 //
-class JvmtiVTMSTransitionDisabler {
+class JvmtiVTMSTransitionDisabler : public AnyObj {
```
> Why this change?
To be more clear about the change:
We need JvmtiVTMSTransitionDisabler for virtual threads only, for platform threads handshake is enough. And currently for platform threads JvmtiVTMSTransitionDisabler disables all mount/unmount transitions (and I'd prefer to avoid this).
So we cannot create JvmtiVTMSTransitionDisabler on a stack (this is standard use scenario), but need to create it conditionally by `new`/`delete`

dholmes-ora replied:
So `JvmtiVTMSTransitionDisabler` is no longer a RAII helper type? That sounds a backward step in usability.

alexmenkov replied:
No. Sorry, my comment had a error (I updated it). This change just allow `JvmtiVTMSTransitionDisabler` to be heap allocated.
Class `TransitionDisabler` in `VMThreadSnapshot::get_thread_snapshot` (line 1444 in threadService.cpp) is kind of smart pointer to `JvmtiVTMSTransitionDisabler`

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+    }
+  };
+
+  Handle _java_thread;
```
I'm going to rename the fields as Serguei suggested (so this will be `Handle _thread_h`)

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+    _blocker_owner.release(oop_storage());
+  }
+
+  bool read_reset_retry() {
```
read and then reset the `retry` value (return old value)

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+
+  Handle _java_thread;
+  JavaThread* _thread;
+  int _depth;
```
fixed

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+  OopHandle _name;
+  GrowableArray<OwnedLock>* _locks;
+  Blocker _blocker;
+  OopHandle _blocker_owner;
```
after offline discussion agreed that in most cases additional explanations are not required

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+    _methods(nullptr), _bcis(nullptr),
+    _thread_status(), _name(nullptr),
+    _locks(nullptr), _blocker(), _blocker_owner(nullptr) { }
+  virtual ~GetThreadSnapshotClosure() {
```
Fixed

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+  bool read_reset_retry() {
+    bool ret = _retry_handshake;
+    // If we re-execute the handshake this method need to return false
+    // when the handshake cannot be performed. (E.g. thread terminating)
```
dropped "retry" logic

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1119,430 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+      return;
+    }
+
+    bool walk_cont = (_java_thread != nullptr) && (_java_thread->vthread_continuation() != nullptr);
```
The logic is the same as before.
The problem is the variable name. it's `bool vthread_carrier` argument of vframeStream ctor. Will change the name

AlanBateman replied:
Something isn't right here. Are you sure walk_cont is correctly for the mounted virtual thread case? There are several tests failing now.

AlanBateman replied:
If renamed to bool vthread_carrier then it will be clearer to see that the logic isn't right. It should only be true when !is_virtual and there is a mounted virtual thread.

alexmenkov replied:
Right, thanks for catching. Fixed

sspitsyn commented:
Alex, thank you for the updates! It is getting better. :)

kevinjwalls commented on src/hotspot/share/services/threadService.hpp:
```diff
@@ -631,4 +631,11 @@ class JavaThreadSleepState : public JavaThreadStatusChanger {
   }
 };
 
+
```
I'm finding the class name VMThreadSnapshot confusing, as it's used to create a snapshot of any thread in the VM, not of the "VMThread".  This space is crowded, we have a class ThreadSnapshot here already.

VMThreadSnapshot is an AllStatic class with one method for creating snapshots... How about ThreadSnapshotFactory?

alexmenkov replied:
Renamed to ThreadSnapshotFactory

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1119,430 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+      ELIMINATED = 1,
+    };
+
+    int _depth;
```
done

alexmenkov commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -1115,3 +1119,430 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+  GrowableArray<Method*>* _methods;
+  GrowableArray<int>* _bcis;
+  JavaThreadStatus _thread_status;
+  OopHandle _name;
```
yes, forgot about it. done

kevinjwalls commented:
Just to be clear, the ifdef JVMTI means Thread.dump_to_file not work in a build configuration using - --enable-jvm-feature-jvmti=no ?
Currently, Thread.dump_to_file works in such configs.  
I don't think this is a problem.. just thought we could spell it out.
This all looks good to me.

AlanBateman commented:
> Just to be clear, the ifdef JVMTI means Thread.dump_to_file not work in a build configuration using - --enable-jvm-feature-jvmti=no ? Currently, Thread.dump_to_file works in such configs. I don't think this is a problem.. just thought we could spell it out. This all looks good to me.

The dependency on JvmtiVTMSTransitionDisabler is needed now but maybe in the future this could be separated from JVMTI for other cases where transitions needs to be disabled.

Minimal VM disables management and jvmti, and is really only practical if you run jlink to create a run-time image with a small subset of the modules. So no jdk.management and no HotSpotDiagnosticMXBean. It's possible of course that someone configures with +management and -jvmti, in which case the HotSpotDiagnosticMXBean.dumpThreads will throw NPE. It should throw UOE (the method is specified to allow this), which means JVM_CreateThreadSnapshot or ThreadSnapshot.of(Thread) should throw.

kevinjwalls commented:
Thanks Alan, and thanks Alex!
